### ISSUE TITLE
Extend rclone serve webdav to support the owncloud extensions for modtime and checksums

### ISSUE THREAD

**Issue by jcasale** [rclone/rclone#3147]
#### What is your current rclone version (output from `rclone version`)?
1.47

#### What problem are you are trying to solve?
The native webdav implementation does not support modtime and checksums when synchronizing. As a result, the reliability of the transfer is compromised. The issue is highly visible when syncing git repos, after the initial transfer, all subsequent syncs are invalid.

#### How do you think rclone should be changed to solve that?
Extend `rclone serve webdav` to support the owncloud extensions for modtime and checksums.

**Comment by ncw** [comment]
This should be relatively straight forward to implement the owncloud extensions.

The first is the modtime for the uploaded file is delivered in a header like this

```
	if f.useOCMtime {
		opts.ExtraHeaders["X-OC-Mtime"] = fmt.Sprintf("%f", float64(src.ModTime().UnixNano())/1E9)
	}
```

And the second is to return the checksum of the file in some extra properties the `oc:checksums` here.  This should be possible too - the webdav server library is capable of adding custom properties.

```xml
<d:propfind  xmlns:d="DAV:" xmlns:oc="http://owncloud.org/ns" xmlns:nc="http://nextcloud.org/ns">
 <d:prop>
  <d:displayname />
  <d:getlastmodified />
  <d:getcontentlength />
  <d:resourcetype />
  <d:getcontenttype />
  <oc:checksums />
 </d:prop>
</d:propfind>
```

**Comment by gary-kim** [comment]
I'm working on the FTPS support right now but if noone's taken this by the time that gets merged, I'll also have a go at this.

**Comment by ncw** [comment]
> I'm working on the FTPS support right now but if noone's taken this by the time that gets merged, I'll also have a go at this.

That would be super :-)  Let me know if you need any help.  I'm not 100% sure how easy or not it will be.

**Comment by gary-kim** [comment]
Also, I noticed while looking at this, is go1.8 support being officially dropped? If so, we can change the `"context"` imports.

**Comment by ncw** [comment]
> Also, I noticed while looking at this, is go1.8 support being officially dropped? If so, we can change the `"context"` imports.

Looks like I missed some things when deprecating go1.8! I'll patch these shortly as there are quite a few more than just `serve/webdav`!

**Comment by ncw** [comment]
This has been implemented already so I'll close this issue.

**Comment by ncw** [comment]
I closed this by accident as was pointed out [on the forum](https://forum.rclone.org/t/rclone-serve-webdav-to-rclone-webdav-not-working-with-modtime-and-hashes-as-expected-from-closed-issue-3147/21882).

Does anyone want to have a go at this?

**Comment by WeidiDeng** [comment]
When using rclone to sync, rclone always propfind the whole directory, that's listing. So checksum caching needs to be enabled. That's also why I decided to store them in my own implementation. I don't know about oc/nc clients though.

md5 and sha1 indeed cover most of the hashes that's used. But I found some quirks using rclone. When configuring a webdav server as owncloud, rclone uses md5 as hash method and won't fallback to sha1 as necessary. (Owncloud does save its own checksum and using client header as verification). So my implementation calculate checksum as body is read.

**Comment by ncw** [comment]
The support for setting modtimes has now been merged - thank you @WeidiDeng 

I've merged this to master now which means it will be in [the latest beta](https://beta.rclone.org/) in 15-30 minutes and released in v1.63

I think this issue is done as we now have modtime and checksum support for owncloud and nextcloud - would you agree @WeidiDeng ?

**Comment by WeidiDeng** [comment]
Mostly done. I'm working on a patch and see if that works well enough.

**Comment by Jwink3101** [comment]
Given that #6990 was merged, this should be closed

**Comment by ncw** [comment]
This will be released in 1.63 - thanks @WeidiDeng
